### PR TITLE
feat(acns): add advanced network security fqdn policy feature to az aks create/update

### PR DESCRIPTION
**Pull request by jshr-w**
Add azcli commands for enabling Advanced Network Security FQDN policy feature for aks-preview extension.
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### Related command

`az aks create --enable-fqdn-policy`
`az aks update --enable-fqdn-policy`
`az aks update --disable-fqdn-policy`

### General Guidelines

- [X] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [X] Have you run `python scripts/ci/test_index.py -q` locally? (`pip install wheel==0.30.0` required)
- [X] My extension version conforms to the [Extension version schema](https://github.com/Azure/azure-cli/blob/release/doc/extensions/versioning_guidelines.md)

For new extensions:

- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your pull request is merged into main branch, a new pull request will be created to update `src/index.json` automatically.  
You only need to update the version information in file setup.py and historical information in file HISTORY.rst in your PR but do not modify `src/index.json`.

### CODE REVIEW

**Comment by yonzhan** [comment]
Thank you for your contribution! We will review the pull request and get back to you soon.

File: src/aks-preview/azext_aks_preview/_help.py [hunk 1]
```diff
@@ -1228,6 +1228,12 @@
         - name: --disable-advanced-network-observability
           type: bool
           short-summary: Disable advanced network observability functionalities on a cluster
+        - name: --enable-fqdn-policy
```
Comment (FumingZhang) [inline src/aks-preview/azext_aks_preview/_help.py:1231]: This seems to be the help msg for `az aks update`, please also add the info for `az aks create`.

File: src/aks-preview/azext_aks_preview/managed_cluster_decorator.py [hunk 2]
```diff
@@ -3024,6 +3044,16 @@ def set_up_network_profile(self, mc: ManagedCluster) -> ManagedCluster:
...
                 )
             )
+
+        fqdn_policy = self.context.get_enable_fqdn_policy()
+        if fqdn_policy is not None:
+            network_profile.advanced_networking = self.models.AdvancedNetworking(  # pylint: disable=no-member
+                security=self.models.AdvancedNetworkingSecurity(  # pylint: disable=no-member
+                    fqdn_policy=self.models.AdvancedNetworkingFQDNPolicy(
+                        enabled=fqdn_policy
+                    )
```
Comment (FumingZhang) [inline src/aks-preview/azext_aks_preview/managed_cluster_decorator.py:3054]: Can observability and security be enabled at the same time? If so, you need to rewrite this to make sure it is empty (None/null) before instantiating the AdvancedNetworking model.

File: src/aks-preview/azext_aks_preview/managed_cluster_decorator.py [hunk 3]
```diff
@@ -4129,6 +4159,24 @@ def update_enable_advanced_network_observability_in_network_profile(self, mc: Ma
...
+    def update_enable_fqdn_policy_in_network_profile(self, mc: ManagedCluster) -> ManagedCluster:
+        """Update enable fqdn policy of network profile for the ManagedCluster object.
+
+        :return: the ManagedCluster object
+        """
+        self._ensure_mc(mc)
+
+        fqdn_policy = self.context.get_enable_fqdn_policy()
+        if fqdn_policy is not None:
+            mc.network_profile.advanced_networking = self.models.AdvancedNetworking(  # pylint: disable=no-member
```
Comment (FumingZhang) [inline src/aks-preview/azext_aks_preview/managed_cluster_decorator.py:4171]: This may overwrite the entire AdvancedNetworking model

**Review by FumingZhang — COMMENTED** [review]
Queued [live test](https://dev.azure.com/msazure/CloudNativeCompute/_build/results?buildId=99890605&view=results) to validate the change.

- test_aks_create_with_enable_fqdn_policy
- test_aks_update_enable_fqdn_policy

File: src/aks-preview/azext_aks_preview/_help.py (hunk 1 above)
Comment (jshr-w) [reply src/aks-preview/azext_aks_preview/_help.py:1231]: Thanks for the catch!

File: src/aks-preview/azext_aks_preview/managed_cluster_decorator.py (hunk 2 above)
Comment (jshr-w) [reply src/aks-preview/azext_aks_preview/managed_cluster_decorator.py]: Thanks, appreciate the pointer!

**Comment by FumingZhang** [comment]
> Queued [live test](https://dev.azure.com/msazure/CloudNativeCompute/_build/results?buildId=99890605&view=results) to validate the change.
> 
> * test_aks_create_with_enable_fqdn_policy
> * test_aks_update_enable_fqdn_policy

The previous live test failed, seems the feature is protected by a feature flag or toggle? Is the feature available in all regions? If it supports bypassing validation via akshttpcustomfeatures header, please modify the test accordingly.

Re-queued [live test](https://dev.azure.com/msazure/CloudNativeCompute/_build/results?buildId=100107782&view=results) in staging env.

File: src/aks-preview/azext_aks_preview/_params.py [hunk 4]
```diff
@@ -1312,8 +1317,18 @@ def load_arguments(self, _):
...
+            action="store_true",
+            is_preview=True,
+        )
+        c.argument(
+            "disable_fqdn_policy",
+            action="store_true",
+            is_preview=True,
+        )
+        c.argument("enable_cost_analysis", is_preview=True, action="store_true")
+        c.argument("disable_cost_analysis", is_preview=True, action="store_true")
```
Comment (FumingZhang) [inline src/aks-preview/azext_aks_preview/_params.py:1331]: seems there is a problem on how you handled merge conflict, `is_preview=True` is removed in PR #7851

**Comment by FumingZhang** [comment]
> The toggle to release feature is currently rolling out, should get to eastus2euap by tomorrow.

If the feature availability is controlled by toggle, there is a problem that it cannot be executed _smoothly_ by the cli runner (something similar to runner, run tests at fixed intervals in a region in prod sub), please exclude your newly added cases in the following configuration file `src/aks-preview/azcli_aks_live_test/configs/ext_matrix_default.json`

File: src/aks-preview/azext_aks_preview/_params.py (hunk 4 above)
Comment (jshr-w) [reply src/aks-preview/azext_aks_preview/_params.py]: Thanks for the catch

**Comment by jshr-w** [comment]
Disabled the test cases until feature rolls out fully to prod.

**Comment by jshr-w** [comment]
[Here](https://dev.azure.com/msazure/CloudNativeCompute/_build/results?buildId=100458801&view=results) is the passing tests run in staging. I have added all tests to the ext_matrix to be excluded from the pipeline until the feature is fully rolled out to all regions.

Tests:
- test_aks_create_with_enable_fqdn_policy 
- test_aks_update_enable_fqdn_policy 
- test_aks_update_enable_fqdn_policy 
- test_aks_create_with_enable_acns 
- test_aks_create_with_enable_acns_complex 
- test_aks_update_enable_acns

**Review by zhoxing-ms — APPROVED** [review]

**Comment by zhoxing-ms** [comment]
@FumingZhang Could you please help review it again?

**Pull request merged**